Multi-project library and transitive dependencies

## Expected Behavior

This is similar to https://github.com/clojurephant/clojurephant/issues/109

In a multi-project build I have an application A that depends on a sibling library B. B has a few third party dependencies. When compileClojure runs on  A, B's third party libraries should be on the classpath.

## Current Behavior

Currently, I see B (`franz` in the logs) compile fist, as expected but when A (`galileo`) compiles it throws an exception that it can't find a dependency of `franz` (`org.apache.kafka.common.TopicPartition`)

```

> Task :cljbeat:franz:checkClojure
Warning: protocol #'cb.cljbeat.franz/ManualCommittingFranzMachine is overwriting method init of protocol AutocommittingFranzMachine
Warning: protocol #'cb.cljbeat.franz/ManualCommittingFranzMachine is overwriting method step of protocol AutocommittingFranzMachine

> Task :cljbeat:franz:compileClojure
Warning: protocol #'cb.cljbeat.franz/ManualCommittingFranzMachine is overwriting method init of protocol AutocommittingFranzMachine
Warning: protocol #'cb.cljbeat.franz/ManualCommittingFranzMachine is overwriting method step of protocol AutocommittingFranzMachine

> Task :astronomy:galileo:checkClojure FAILED
ERROR Syntax error compiling at (cb/galileo/pipeline.clj:1:1).
Exception in thread "main" Syntax error compiling at (cb/galileo/pipeline.clj:1:1).
        at clojure.lang.Compiler.load(Compiler.java:7647)
        at clojure.lang.RT.loadResourceScript(RT.java:381)
        at clojure.lang.RT.loadResourceScript(RT.java:372)
        at clojure.lang.RT.load(RT.java:463)
        at clojure.lang.RT.load(RT.java:428)
        at clojure.core$load$fn__6824.invoke(core.clj:6126)
        at clojure.core$load.invokeStatic(core.clj:6125)
        at clojure.core$load.doInvoke(core.clj:6109)
        at clojure.lang.RestFn.invoke(RestFn.java:408)
        at dev.clojurephant.tools.clojure_loader$_main$fn__174.invoke(clojure_loader.clj:30)
        at dev.clojurephant.tools.clojure_loader$_main.invokeStatic(clojure_loader.clj:23)
        at dev.clojurephant.tools.clojure_loader$_main.doInvoke(clojure_loader.clj:19)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at clojure.lang.Var.applyTo(Var.java:705)
        at clojure.core$apply.invokeStatic(core.clj:665)
        at clojure.main$main_opt.invokeStatic(main.clj:491)
        at clojure.main$main_opt.invoke(main.clj:487)
        at clojure.main$main.invokeStatic(main.clj:598)
        at clojure.main$main.doInvoke(main.clj:561)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.lang.Var.applyTo(Var.java:705)
        at clojure.main.main(main.java:37)
Caused by: java.lang.ExceptionInInitializerError
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at clojure.lang.RT.classForName(RT.java:2207)
        at clojure.lang.RT.classForName(RT.java:2216)
        at clojure.lang.RT.loadClassForName(RT.java:2235)
        at clojure.lang.RT.load(RT.java:453)
        at clojure.lang.RT.load(RT.java:428)
        at clojure.core$load$fn__6824.invoke(core.clj:6126)
        at clojure.core$load.invokeStatic(core.clj:6125)
        at clojure.core$load.doInvoke(core.clj:6109)
        at clojure.lang.RestFn.invoke(RestFn.java:408)
        at clojure.core$load_one.invokeStatic(core.clj:5908)
        at clojure.core$load_one.invoke(core.clj:5903)
        at clojure.core$load_lib$fn__6765.invoke(core.clj:5948)
        at clojure.core$load_lib.invokeStatic(core.clj:5947)
        at clojure.core$load_lib.doInvoke(core.clj:5928)
        at clojure.lang.RestFn.applyTo(RestFn.java:142)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$load_libs.invokeStatic(core.clj:5985)
        at clojure.core$load_libs.doInvoke(core.clj:5969)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$require.invokeStatic(core.clj:6007)
        at clojure.core$require.doInvoke(core.clj:6007)
        at clojure.lang.RestFn.invoke(RestFn.java:512)
        at cb.cljbeat.franz$loading__6706__auto____176.invoke(franz.clj:1)
        at cb.cljbeat.franz__init.load(Unknown Source)
        at cb.cljbeat.franz__init.<clinit>(Unknown Source)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at clojure.lang.RT.classForName(RT.java:2207)
        at clojure.lang.RT.classForName(RT.java:2216)
        at clojure.lang.RT.loadClassForName(RT.java:2235)
        at clojure.lang.RT.load(RT.java:453)
        at clojure.lang.RT.load(RT.java:428)
        at clojure.core$load$fn__6824.invoke(core.clj:6126)
        at clojure.core$load.invokeStatic(core.clj:6125)
        at clojure.core$load.doInvoke(core.clj:6109)
        at clojure.lang.RestFn.invoke(RestFn.java:408)
        at clojure.core$load_one.invokeStatic(core.clj:5908)
        at clojure.core$load_one.invoke(core.clj:5903)
        at clojure.core$load_lib$fn__6765.invoke(core.clj:5948)
        at clojure.core$load_lib.invokeStatic(core.clj:5947)
        at clojure.core$load_lib.doInvoke(core.clj:5928)
        at clojure.lang.RestFn.applyTo(RestFn.java:142)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$load_libs.invokeStatic(core.clj:5985)
        at clojure.core$load_libs.doInvoke(core.clj:5969)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$require.invokeStatic(core.clj:6007)
        at clojure.core$require.doInvoke(core.clj:6007)
        at clojure.lang.RestFn.invoke(RestFn.java:805)
        at cb.galileo.pipeline$eval188$loading__6706__auto____189.invoke(pipeline.clj:1)
        at cb.galileo.pipeline$eval188.invokeStatic(pipeline.clj:1)
        at cb.galileo.pipeline$eval188.invoke(pipeline.clj:1)
        at clojure.lang.Compiler.eval(Compiler.java:7176)
        at clojure.lang.Compiler.eval(Compiler.java:7165)
        at clojure.lang.Compiler.load(Compiler.java:7635)
        ... 23 more
Caused by: java.lang.ClassNotFoundException: org.apache.kafka.common.TopicPartition
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at clojure.lang.RT.classForName(RT.java:2207)
        at clojure.lang.RT.classForNameNonLoading(RT.java:2220)
        at cb.cljbeat.franz.consumer$loading__6706__auto____178.invoke(consumer.clj:1)
        at cb.cljbeat.franz.consumer__init.load(Unknown Source)
        at cb.cljbeat.franz.consumer__init.<clinit>(Unknown Source)
        ... 82 more
```

I tried doing the same but using the java-library plugin and setting the dependencies to `api` (they are currently `implementation`) but that failed to put the code for my library on the application classpath at all. 

The deps for the application looks like this

```
dependencies {
  implementation commonDeps.cljcore
  implementation commonDeps.logging
  implementation commonDeps.component
  implementation commonDeps.cbcore
  implementation commonDeps.core_cache
  implementation 'org.flatland:ordered:1.5.7'
  implementation project(':cljbeat:franz')
  implementation commonDeps.msgpack
  implementation cbDeps.pingdecoder
  devImplementation commonDeps.devtools
  testImplementation 'junit:junit:4.12'
}
```

and for the library 

```
dependencies {
  implementation commonDeps.cljcore
  implementation 'clojurewerkz:propertied:1.3.0'
  implementation 'org.apache.kafka:kafka-clients:0.10.2.0'
  devImplementation commonDeps.devtools
  testImplementation 'junit:junit:4.12'
}
```

Thanks! This is a great project. Not having a reliable way to manage multi-project builds in clojure is a major issue for us as our code base grows. 

- **clojurephant version:** 0.5.0-alpha.5
- **Gradle version:** 5.4.1
- **Java version:** 1.8
- **OS version:** OSX

--- a/clojurephant-plugin/src/compatTest/clojure/dev/clojurephant/compat_test/multi_project.clj
+++ b/clojurephant-plugin/src/compatTest/clojure/dev/clojurephant/compat_test/multi_project.clj
@@ -1,0 +1,20 @@
+(ns dev.clojurephant.compat-test.multi-project
+  (:require [clojure.set :as set]
+            [clojure.string :as str]
+            [clojure.test :refer :all]
+            [dev.clojurephant.compat-test.test-kit :as gradle]
+            [ike.cljj.file :as file])
+  (:import [org.gradle.testkit.runner TaskOutcome]))
+
+(deftest multi-project-classpath
+  (testing "with multi-project and AOT off, project B sources and dependencies are on classpath of A"
+    (gradle/with-project "MultiProjectTest"
+      (let [result (gradle/build "jar")]
+        (gradle/verify-task-outcome result ":SubClojureB:compileClojure" :skipped)
+        (gradle/verify-task-outcome result ":SubClojureB:checkClojure" :success)
+        (gradle/verify-task-outcome result ":SubClojureB:jar" :success)
+        (gradle/verify-task-outcome result ":SubClojureA:compileClojure" :skipped)
+        (gradle/verify-task-outcome result ":SubClojureA:checkClojure" :success)
+        (gradle/verify-task-outcome result ":SubClojureA:jar" :success)
+        (gradle/verify-jar-contents ["SubClojureA/src/main/clojure" "SubClojureA/src/main/resources"] "SubClojureA/build/libs/SubClojureA.jar")
+        (gradle/verify-jar-contents ["SubClojureB/src/main/clojure" "SubClojureB/src/main/resources"] "SubClojureB/build/libs/SubClojureB.jar")))))

--- a/clojurephant-plugin/src/compatTest/clojure/dev/clojurephant/compat_test/test_kit.clj
+++ b/clojurephant-plugin/src/compatTest/clojure/dev/clojurephant/compat_test/test_kit.clj
@@ -33,7 +33,9 @@
         xf (comp (drop 1)
                  (filter file/file?)
                  (map (fn [f] (.relativize root f))))]
-    (into #{} xf (file/walk root))))
+    (if (file/exists? root)
+      (into #{} xf (file/walk root))
+      #{})))
 
 (defn verify-compilation-without-aot
   [src-dir dst-dir]

--- a/clojurephant-plugin/src/compatTest/projects/MultiProjectTest/SubClojureA/build.gradle
+++ b/clojurephant-plugin/src/compatTest/projects/MultiProjectTest/SubClojureA/build.gradle
@@ -1,0 +1,9 @@
+plugins {
+  id 'dev.clojurephant.clojure'
+  id 'java-library'
+}
+
+dependencies {
+  api 'org.clojure:clojure:1.8.0'
+  api project(':SubClojureB')
+}

--- a/clojurephant-plugin/src/compatTest/projects/MultiProjectTest/SubClojureA/src/main/clojure/clja/core.clj
+++ b/clojurephant-plugin/src/compatTest/projects/MultiProjectTest/SubClojureA/src/main/clojure/clja/core.clj
@@ -1,0 +1,5 @@
+(ns clja.core
+  (:require [cljb.core :as cljb]))
+
+(defn process []
+  (:my-key (cljb/read-data)))

--- a/clojurephant-plugin/src/compatTest/projects/MultiProjectTest/SubClojureB/build.gradle
+++ b/clojurephant-plugin/src/compatTest/projects/MultiProjectTest/SubClojureB/build.gradle
@@ -1,0 +1,9 @@
+plugins {
+  id 'dev.clojurephant.clojure'
+  id 'java-library'
+}
+
+dependencies {
+  api 'org.clojure:clojure:1.8.0'
+  api 'org.clojure:data.json:1.0.0'
+}

--- a/clojurephant-plugin/src/compatTest/projects/MultiProjectTest/SubClojureB/src/main/clojure/cljb/core.clj
+++ b/clojurephant-plugin/src/compatTest/projects/MultiProjectTest/SubClojureB/src/main/clojure/cljb/core.clj
@@ -1,0 +1,5 @@
+(ns cljb.core
+  (:require [clojure.data.json :as json]))
+
+(defn read-data []
+  (json/read-str "{ \"my-key\": 123 }"))

--- a/clojurephant-plugin/src/compatTest/projects/MultiProjectTest/build.gradle
+++ b/clojurephant-plugin/src/compatTest/projects/MultiProjectTest/build.gradle
@@ -1,0 +1,12 @@
+subprojects {
+  plugins.withId('dev.clojurephant.clojure') {
+    repositories {
+      mavenCentral()
+      maven {
+        name = 'Clojars'
+        url = 'https://repo.clojars.org/'
+      }
+      mavenLocal()
+    }
+  }
+}

--- a/clojurephant-plugin/src/compatTest/projects/MultiProjectTest/settings.gradle
+++ b/clojurephant-plugin/src/compatTest/projects/MultiProjectTest/settings.gradle
@@ -1,0 +1,3 @@
+include 'SubClojureA'
+include 'SubClojureB'
+include 'SubJavaC'

--- a/clojurephant-plugin/src/main/java/dev/clojurephant/plugin/clojure/ClojureBasePlugin.java
+++ b/clojurephant-plugin/src/main/java/dev/clojurephant/plugin/clojure/ClojureBasePlugin.java
@@ -45,17 +45,18 @@ public class ClojureBasePlugin implements Plugin<Project> {
 
       ClojureBuild build = extension.getBuilds().create(sourceSet.getName());
       build.getSourceSet().set(sourceSet);
-      ((DefaultSourceSetOutput) sourceSet.getOutput()).addClassesDir(() -> build.getOutputDir().get().getAsFile());
-      project.getTasks().getByName(sourceSet.getClassesTaskName()).dependsOn(build.getTaskName("compile"));
-      project.getTasks().getByName(sourceSet.getClassesTaskName()).dependsOn(build.getTaskName("check"));
 
-      sourceSet.getOutput().dir(project.provider(() -> {
+      Provider<FileCollection> output = project.provider(() -> {
         if (build.isCompilerConfigured()) {
-          return build.getOutputDir();
+          return project.files(build.getOutputDir());
         } else {
           return clojureSourceSet.getClojure().getSourceDirectories();
         }
-      }));
+      });
+
+      ((DefaultSourceSetOutput) sourceSet.getOutput()).getClassesDirs().from(output);
+      project.getTasks().getByName(sourceSet.getClassesTaskName()).dependsOn(build.getTaskName("compile"));
+      project.getTasks().getByName(sourceSet.getClassesTaskName()).dependsOn(build.getTaskName("check"));
     });
   }
 

--- a/clojurephant-plugin/src/main/java/dev/clojurephant/plugin/clojurescript/ClojureScriptBasePlugin.java
+++ b/clojurephant-plugin/src/main/java/dev/clojurephant/plugin/clojurescript/ClojureScriptBasePlugin.java
@@ -8,9 +8,12 @@ import dev.clojurephant.plugin.clojurescript.tasks.ClojureScriptSourceSet;
 import dev.clojurephant.plugin.common.internal.ClojureCommonBasePlugin;
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
+import org.gradle.api.file.FileCollection;
 import org.gradle.api.internal.plugins.DslObject;
+import org.gradle.api.internal.tasks.DefaultSourceSetOutput;
 import org.gradle.api.model.ObjectFactory;
 import org.gradle.api.plugins.JavaPluginConvention;
+import org.gradle.api.provider.Provider;
 import org.gradle.api.tasks.SourceSet;
 
 public class ClojureScriptBasePlugin implements Plugin<Project> {
@@ -41,16 +44,18 @@ public class ClojureScriptBasePlugin implements Plugin<Project> {
 
       ClojureScriptBuild build = extension.getBuilds().create(sourceSet.getName());
       build.getSourceSet().set(sourceSet);
-      sourceSet.getOutput().dir(build.getOutputDir());
-      project.getTasks().getByName(sourceSet.getClassesTaskName()).dependsOn(build.getTaskName("compile"));
 
-      sourceSet.getOutput().dir(project.provider(() -> {
+      Provider<FileCollection> output = project.provider(() -> {
         if (build.isCompilerConfigured()) {
-          return build.getOutputDir();
+          return project.files(build.getOutputDir());
         } else {
           return clojurescriptSourceSet.getClojureScript().getSourceDirectories();
         }
-      }));
+      });
+
+      ((DefaultSourceSetOutput) sourceSet.getOutput()).getClassesDirs().from(output);
+
+      project.getTasks().getByName(sourceSet.getClassesTaskName()).dependsOn(build.getTaskName("compile"));
     });
   }
 

